drafts/bitcartel-nu0-network-handshake/draft.rst

```diff
@@ -0,0 +1,237 @@
+::
+
+  ZIP: 201
+  Title: Network Peer Management for Overwinter
+  Author: Simon Liu <simon@z.cash>
+  Category: Network
+  Created: 2018-01-15
+  License: MIT
+
+Terminology
+===========
+
+The key words "MUST", "MUST NOT", "SHOULD", and "MAY" in this document are to be interpreted as described in RFC 2119. [#RFC2119]_
+
+The terms below are to be interpreted as follows:
+
+Branch
+  A chain of blocks with common consensus rules, where the first block in the chain is either the genesis
+  block, or the child of a parent block created under an older set of consensus rules (i.e. the parent block
+  is a member of a different branch). By definition, every block belongs to at most one branch.
+
+Hard fork
+  The creation of a new branch by a change in the consensus rules of the network. Nodes that do not recognize
+  the new rules will continue to follow the old branch.
+
+Network upgrade
+  An intentional hard fork undertaken by the community in order to improve the network.
+
+Overwinter
+  Code-name for the first ZCash network upgrade, also known as Network Upgrade Zero.
+
+
+Abstract
+========
+
+This proposal defines an upgrade to the network handshake and peer management required for network upgrades following the Network Upgrade Activation Mechanism [#zip-0200]_.
+
+Related to:
+
+- Transaction Signature Verification for Overwinter [#zip-0143]_.
+- Version 3 Transaction Format for Overwinter [#zip-0202]_.
+- Transaction Expiry [#zip-0203]_.
+
+Motivation
+==========
+
+With scheduled network upgrades, at the activation height, nodes on each branch should disconnect from nodes on other branches and only accept new incoming connections from nodes on the same branch.
+
+Specification
+=============
+
+When a new inbound connection is received or an outbound connection created, a CNode object is instantiated with the version field set to INIT_PROTO_VERSION which has a value of 209. This value is not transmitted across the network, but for legacy reasons and technical debt beyond the scope of this ZIP, this value will not be changed.
+
+Once the two nodes have connected and started the handshake to negotiate the protocol version, the version field of CNode will be updated.  The handshake [#bitcoin-version-handshake]_ involves "version" and "verack" messages being exchanged.::
+
+    L -> R: Send version message with the local peer's version
+    R -> L: Send version message back
+    R -> L: Send verack message
+    R:      Sets version to the minimum of the 2 versions
+    L -> R: Send verack message after receiving version message from R
+    L:      Sets version to the minimum of the 2 versions
+
+To send a version message, the node will invoke ``PushVersion()``::
+
+    void CNode::PushVersion() {
+        ...
+        PushMessage("version", PROTOCOL_VERSION, nLocalServices, nTime, addrYou, addrMe, ...);
+        ...
+    }
+      
+where:
+
+- Pre-Overwinter PROTOCOL_VERSION is 170002
+- OVERWINTER_PROTO_VERSION is 170003
```

Comment:
Replace the above four lines with:

```
where ``PROTOCOL_VERSION`` is the highest protocol version supported by the node.
```

We should also try to make this document Overwinter-version-agnostic, since there will be two protocol versions associated with Overwinter.